[master] Provide block number to Scilla via IPC instead of blockchain JSON

vaivaswatha opened the pull request:
References:
https://github.com/Zilliqa/scilla/pull/1056
https://github.com/Zilliqa/scilla-compiler/wiki/Interaction-with-Blockchain#blockchain-queries

## Backward Compatibility
<!-- For breaking changes, code must be protected by UPGRADE_TARGET -->
- [x] This is not a breaking change
- [ ] This is a breaking change

## Review Suggestion
<!-- How should the reviewers get started on reviewing your pull request-->
<!-- How can the reviewers verify the pull request is working as expected -->

## Status

### Implementation
<!-- Add more TODOs before "ready for review", if any  -->
- [x] **ready for review**

### Integration Test (Core Team)
Existing unit tests cover this.
<!-- This is for core team only, ignore this if you are a community contributor -->

<!-- append the commit digest to inform the others of the versions you have tested -->
- [x] local machine test
<!-- - [ ] local machine test (commit: bbbbbbbb) -->
<!-- - [ ] local machine test (commit: cccccccc) -->
<!-- - [ ] local machine test (commit: dddddddd) -->
- [ ] small-scale cloud test
<!-- - [ ] small-scale cloud test (commit: bbbbbbbb) -->
<!-- - [ ] small-scale cloud test (commit: cccccccc) -->
<!-- - [ ] small-scale cloud test (commit: dddddddd) -->

sandipbhoir commented on src/libData/AccountData/AccountStoreSC.h:
```diff
@@ -24,10 +24,30 @@
+        m_rootHash(rootHash),
+        m_scillaVersion(scillaVersion) {}
+
+  uint64_t getCurBlockNum() const { return m_curBlockNum; }
```
May consider changing return types to `const &` for these getter function?

sandipbhoir commented:
@vaivaswatha seems to have travis build ( test ) failure.

ansnunez commented on src/libData/AccountData/AccountStore.cpp:
```diff
@@ -817,8 +817,11 @@ bool AccountStore::MigrateContractStates(
+    // prepare IPC with current blockchain info provider.
+    ScillaBCInfoProvider sbcip(getCurBlockNum(), address,
+                               account.GetStorageRoot(), scilla_version);
+    m_scillaIPCServer->setBCInfoProvider(&sbcip);
```
I'm not sure it's completely safe to assign the address of a local variable (`sbcip`) for later use in other functions unless it's guaranteed the local variable always still remains in its current scope. Maybe smart pointer or even pass-by-value would be better, even if less efficient.

vaivaswatha commented:
I've addressed the review comments.

Regarding the CI failure, this PR requires that https://github.com/Zilliqa/scilla/pull/1056 be merged. I'll do that now. Does the CI take latest `master` from Scilla repo?

ansnunez commented:
> I've addressed the review comments.
> 
> Regarding the CI failure, this PR requires that [Zilliqa/scilla#1056](https://github.com/Zilliqa/scilla/pull/1056) be merged. I'll do that now. Does the CI take latest `master` from Scilla repo?

Nope, we take a commit hash ([here](https://github.com/Zilliqa/Zilliqa/blob/3a6c9ff1d4e6cbbf5b230032b470f5c292584703/docker/Dockerfile#L17)). We can update this once the merge on the Scilla repo has been done.

sandipbhoir commented on src/libData/AccountData/AccountStoreSC.h:
```diff
@@ -37,7 +37,7 @@ class ScillaBCInfoProvider : public ScillaBCInfo {
 
-  uint64_t getCurBlockNum() const { return m_curBlockNum; }
+  const uint64_t& getCurBlockNum() const { return m_curBlockNum; }
   dev::h256 getRootHash() const { return m_rootHash; }
```
Sorry May be I wasn’t clear. Can we change it for all rest of 3 getter functions?

ansnunez commented on src/libServer/ScillaIPCServer.h:
```diff
@@ -25,6 +25,15 @@
 #include "libData/AccountData/Account.h"
 
+// An abstract base class to provide various blockchain info to Scilla.
+class ScillaBCInfo {
```
Just curious, is the abstraction because the possibility of different providers exists in the near future? Because if not, maybe we can consider removing the abstraction for now, if we want to squeeze out whatever optimization we can on this front. The code looks good, but the creation of virtual table for this class (with only a single function implementation for each table entry) adds to execution time.

vaivaswatha commented:
@ansnunez Removed the abstraction. I don't recall why I added it, but at the time of doing this change, it was more of a "draft" on what needs to be done and I just assumed that someone will take over. But now that we're reviewing it formally, let's do it right.

@sandipbhoir Done.

>We can update this once the merge on the Scilla repo has been done.
Please update the commit to this: https://github.com/Zilliqa/scilla/commit/47de22b4d95b6bf19a49d80aedc9f6b714938c23
